### ISSUE TITLE
Entry module not found: Error: Can't resolve 'html-webpack-plugin/lib/loader.j  s'

### ISSUE THREAD

**Issue by JanStureNielsen** [angular/angular-cli#4237]
### OS?
> Windows 7, 8 or 10. Linux (which distribution). Mac OSX (Yosemite? El Capitan?)

`Linux 3.16.7 openSUSE 13.2`

### Versions.
> Please run `ng --version`. If there's nothing outputted, please run in a Terminal: `node --version` and paste the result here:
```
jan@linux:~/proj> ng --version                                                                                                  

                             _                           _  _                                                                                      
  __ _  _ __    __ _  _   _ | |  __ _  _ __         ___ | |(_)                                                                                     
 / _` || '_ \  / _` || | | || | / _` || '__|_____  / __|| || |                                                                                     
| (_| || | | || (_| || |_| || || (_| || |  |_____|| (__ | || |                                                                                     
 \__,_||_| |_| \__, | \__,_||_| \__,_||_|          \___||_||_|                                                                                     
               |___/                                                                                                                               
angular-cli: 1.0.0-beta.25.5
node: 6.9.1
os: linux x64
@angular/common: 2.4.5
@angular/compiler: 2.4.5
@angular/core: 2.4.5
@angular/forms: 2.4.5
@angular/http: 2.4.5
@angular/platform-browser: 2.4.5
@angular/platform-browser-dynamic: 2.4.5
@angular/router: 3.4.5
@angular/material: 2.0.0-beta.1
@angular/compiler-cli: 2.4.5
```

### Repro steps.
> Was this an app that wasn't created using the CLI? What change did you do on your code? etc.

I upgraded from `angular-cli@1.0.0-beta.21` to `angular-cli@1.0.0-beta.25.5`, making all necessary changes to get my app to compile and run locally. When I attempt to deploy to Heroku, however, it fails with error below.

This is likely a dependency issue specific to Heroku but after searching through webpack, html-webpack-plugin, and stackoverflow, I have found nothing that resolves this. The spaces in `loader.j  s'` and `build_6a557f  153da6022c883b49ad04b6209f'` look like a problem, but I don't know how to tickle `angular-cli+webpack` to impact that.

Any ideas what I'm missing?

### The log given by the failure.
> Normally this include a stack trace and some more information.
```
remote:        ERROR in   Error: Child compilation failed:
remote:        Entry module not found: Error: Can't resolve 'html-webpack-plugin/lib/loader.j  s' in '/tmp/build_6a557f153da6022c883b49ad04b6209f':
remote:        Error: Can't resolve 'html-webpack-plugin/lib/loader.js' in '/tmp/build_6a557f  153da6022c883b49ad04b6209f'

remote:        - compiler.js:76
remote:        [lib]/[angular-cli]/[html-webpack-plugin]/lib/compiler.js:76:16
remote:        
remote:        - Compiler.js:280 Compiler.<anonymous>
remote:        [lib]/[angular-cli]/[webpack]/lib/Compiler.js:280:10
remote:        
remote:        - Compiler.js:480
remote:        [lib]/[angular-cli]/[webpack]/lib/Compiler.js:480:13
remote:        
remote:        - Tapable.js:138 next
remote:        [lib]/[angular-cli]/[tapable]/lib/Tapable.js:138:11
remote:        
remote:        - CachePlugin.js:62 Compiler.<anonymous>
remote:        [lib]/[angular-cli]/[webpack]/lib/CachePlugin.js:62:5
remote:        
remote:        - Tapable.js:142 Compiler.applyPluginsAsyncSeries
remote:        [lib]/[angular-cli]/[tapable]/lib/Tapable.js:142:13
remote:        
remote:        - Compiler.js:477
remote:        [lib]/[angular-cli]/[webpack]/lib/Compiler.js:477:10
remote:        
remote:        - Tapable.js:138 next
remote:        [lib]/[angular-cli]/[tapable]/lib/Tapable.js:138:11
remote:        
remote:        - suppress-entry-chunks-webpack-plugin.js:38 Compilation.<anonymous>
remote:        [lib]/[angular-cli]/plugins/suppress-entry-chunks-webpack-plugin.js:38:17
remote:        
remote:        - Tapable.js:142 Compilation.applyPluginsAsyncSeries
remote:        [lib]/[angular-cli]/[tapable]/lib/Tapable.js:142:13
remote:        
remote:        - Compilation.js:615 self.applyPluginsAsync.err
remote:        [lib]/[angular-cli]/[webpack]/lib/Compilation.js:615:19
remote:        
remote:        - Tapable.js:131 Compilation.applyPluginsAsyncSeries
remote:        [lib]/[angular-cli]/[tapable]/lib/Tapable.js:131:46
remote:        
remote:        - Compilation.js:606 self.applyPluginsAsync.err
remote:        [lib]/[angular-cli]/[webpack]/lib/Compilation.js:606:11
remote:        
remote:        - Tapable.js:138 next
remote:        [lib]/[angular-cli]/[tapable]/lib/Tapable.js:138:11
remote:        
remote:        - UglifyJsPlugin.js:140 Compilation.compilation.plugin
remote:        [lib]/[angular-cli]/[webpack]/lib/optimize/UglifyJsPlugin.js:140:5
remote:        
remote:        - Tapable.js:142 Compilation.applyPluginsAsyncSeries
remote:        [lib]/[angular-cli]/[tapable]/lib/Tapable.js:142:13
remote:        
remote:        - Compilation.js:601 self.applyPluginsAsync.err
remote:        [lib]/[angular-cli]/[webpack]/lib/Compilation.js:601:10
remote:        
remote:        - Tapable.js:131 Compilation.applyPluginsAsyncSeries
remote:        [lib]/[angular-cli]/[tapable]/lib/Tapable.js:131:46
remote:        
remote:        - Compilation.js:597 sealPart2
remote:        [lib]/[angular-cli]/[webpack]/lib/Compilation.js:597:9
remote:        
remote:        - Tapable.js:131 Compilation.applyPluginsAsyncSeries
remote:        [lib]/[angular-cli]/[tapable]/lib/Tapable.js:131:46
remote:        
remote:        - Compilation.js:545 Compilation.seal
remote:        [lib]/[angular-cli]/[webpack]/lib/Compilation.js:545:8
remote:        
remote:        - Compiler.js:474
remote:        [lib]/[angular-cli]/[webpack]/lib/Compiler.js:474:16
remote:        
remote:        - Tapable.js:225
remote:        [lib]/[angular-cli]/[tapable]/lib/Tapable.js:225:11
remote:        
remote:        - Compilation.js:452 _addModuleChain
remote:        [lib]/[angular-cli]/[webpack]/lib/Compilation.js:452:11
remote:        
remote:        - Compilation.js:345 Compilation.errorAndCallback
remote:        [lib]/[angular-cli]/[webpack]/lib/Compilation.js:345:4
remote:        
remote:        - Compilation.js:362 moduleFactory.create
remote:        [lib]/[angular-cli]/[webpack]/lib/Compilation.js:362:12
remote:        
remote:        - NormalModuleFactory.js:234
remote:        [lib]/[angular-cli]/[webpack]/lib/NormalModuleFactory.js:234:19
remote:        
remote:        - NormalModuleFactory.js:59 onDoneResolving
remote:        [lib]/[angular-cli]/[webpack]/lib/NormalModuleFactory.js:59:20
remote:        
remote:        - NormalModuleFactory.js:126
remote:        [lib]/[angular-cli]/[webpack]/lib/NormalModuleFactory.js:126:20
remote:        
remote:        - async.js:3694
remote:        [lib]/[angular-cli]/[async]/dist/async.js:3694:9
remote:        
remote:        - async.js:356
remote:        [lib]/[angular-cli]/[async]/dist/async.js:356:16
remote:        
remote:        - async.js:934 iteratorCallback
remote:        [lib]/[angular-cli]/[async]/dist/async.js:934:13
remote:        
remote:        - async.js:840
remote:        [lib]/[angular-cli]/[async]/dist/async.js:840:16
remote:        
remote:        - async.js:3691
remote:        [lib]/[angular-cli]/[async]/dist/async.js:3691:13
remote:        
remote:        - async.js:21 apply
remote:        [lib]/[angular-cli]/[async]/dist/async.js:21:25
remote:        
remote:        - async.js:56
remote:        [lib]/[angular-cli]/[async]/dist/async.js:56:12
remote:        
remote:        - async.js:1013
remote:        [lib]/[angular-cli]/[async]/dist/async.js:1013:9
remote:        
remote:        - async.js:356
remote:        [lib]/[angular-cli]/[async]/dist/async.js:356:16
remote:        
remote:        - async.js:934 iteratorCallback
remote:        [lib]/[angular-cli]/[async]/dist/async.js:934:13
remote:        
remote:        - async.js:840
remote:        [lib]/[angular-cli]/[async]/dist/async.js:840:16
remote:        
remote:        - async.js:1010
remote:        [lib]/[angular-cli]/[async]/dist/async.js:1010:13
remote:        
remote:        - NormalModuleFactory.js:262
remote:        [lib]/[angular-cli]/[webpack]/lib/NormalModuleFactory.js:262:19
remote:        
remote:        - Resolver.js:70 onResolved
remote:        [lib]/[angular-cli]/[webpack]/[enhanced-resolve]/lib/Resolver.js:70:11
remote:        
remote:        - createInnerCallback.js:31 loggingCallbackWrapper
remote:        [lib]/[angular-cli]/[webpack]/[enhanced-resolve]/lib/createInnerCallback.js:    31:19
remote:        
remote:        - Resolver.js:138 afterInnerCallback
remote:        [lib]/[angular-cli]/[webpack]/[enhanced-resolve]/lib/Resolver.js:138:10
remote:        
remote:        - createInnerCallback.js:31 loggingCallbackWrapper
remote:        [lib]/[angular-cli]/[webpack]/[enhanced-resolve]/lib/createInnerCallback.js:    31:19
remote:        
remote:        - Tapable.js:181 Resolver.applyPluginsAsyncSeriesBailResult1
remote:        [lib]/[angular-cli]/[tapable]/lib/Tapable.js:181:46
remote:        
remote:        - Resolver.js:125 innerCallback
remote:        [lib]/[angular-cli]/[webpack]/[enhanced-resolve]/lib/Resolver.js:125:19
remote:        
remote:        - createInnerCallback.js:31 loggingCallbackWrapper
remote:        [lib]/[angular-cli]/[webpack]/[enhanced-resolve]/lib/createInnerCallback.js:    31:19
remote:        
remote:        - Tapable.js:283
remote:        [lib]/[angular-cli]/[tapable]/lib/Tapable.js:283:15
remote:        
remote:        - UnsafeCachePlugin.js:39
remote:        [lib]/[angular-cli]/[webpack]/[enhanced-resolve]/lib/UnsafeCachePlugin.js:39    :4
remote:        
remote:        
remote:        
remote:        ERROR in multi ./src/main.ts
remote:        Module not found: Error: Can't resolve '@ngtools/webpack' in '/tmp/build_6a557f153da6022c883b49ad04b6209f'
remote:        @ multi ./src/main.ts
remote:        
remote:        ERROR in multi ./src/app/assets/css/main.scss
remote:        Module not found: Error: Can't resolve 'style-loader' in '/tmp/build_6a557f153da6022c883b49ad04b6209f'
remote:        @ multi ./src/app/assets/css/main.scss
remote:        
remote:        ERROR in   TypeError: Cannot read property 'match' of undefined
remote:        
remote:        - suppress-entry-chunks-webpack-plugin.js:44 filterFn
remote:        [lib]/[angular-cli]/plugins/suppress-entry-chunks-webpack-plugin.js:44:76
remote:        
remote:        - Array.filter
remote:        
remote:        - suppress-entry-chunks-webpack-plugin.js:47 Compilation.<anonymous>
remote:        [lib]/[angular-cli]/plugins/suppress-entry-chunks-webpack-plugin.js:47:59
remote:        
remote:        - Tapable.js:208 Compilation.applyPluginsAsyncWaterfall
remote:        [lib]/[angular-cli]/[tapable]/lib/Tapable.js:208:13
remote:        
remote:        - util.js:16 Compilation.tryCatcher
remote:        [lib]/[angular-cli]/[bluebird]/js/release/util.js:16:23
remote:        
remote:        
remote:        - index.js:636
remote:        [lib]/[angular-cli]/[html-webpack-plugin]/index.js:636:12
remote:        
remote:        - index.js:152
remote:        [lib]/[angular-cli]/[html-webpack-plugin]/index.js:152:16
remote:        
remote:        - util.js:16 tryCatcher
remote:        [lib]/[angular-cli]/[bluebird]/js/release/util.js:16:23
remote:        
remote:        - promise.js:510 Promise._settlePromiseFromHandler
remote:        [lib]/[angular-cli]/[bluebird]/js/release/promise.js:510:31
remote:        
remote:        - promise.js:567 Promise._settlePromise
remote:        [lib]/[angular-cli]/[bluebird]/js/release/promise.js:567:18
remote:        
remote:        - promise.js:612 Promise._settlePromise0
remote:        [lib]/[angular-cli]/[bluebird]/js/release/promise.js:612:10
remote:        
remote:        - promise.js:691 Promise._settlePromises
remote:        [lib]/[angular-cli]/[bluebird]/js/release/promise.js:691:18
remote:        
remote:        - async.js:133 Async._drainQueue
remote:        [lib]/[angular-cli]/[bluebird]/js/release/async.js:133:16
remote:        
remote:        - async.js:143 Async._drainQueues
remote:        [lib]/[angular-cli]/[bluebird]/js/release/async.js:143:10
remote:        
remote:        - async.js:17 Immediate.Async.drainQueues
remote:        [lib]/[angular-cli]/[bluebird]/js/release/async.js:17:14
```

### Mention any other details that might be useful.

> ---------------------------------------------------------------
> Thanks! We'll be in touch soon.

The `package.json` syntax is valid; `depcheck` shows:

```
jan@linux-zd16:~/src/fm-repos/fm-ui> depcheck .                                                                                                    
Unused dependencies
* ts-helpers
Unused devDependencies
* @angular/compiler-cli
* @types/hammerjs
* @types/jasmine
* @types/node
* codelyzer
* karma-cli
Missing dependencies
* pushstate-server
```

Additional details can be found in [my SO question](http://stackoverflow.com/questions/41877754/entry-module-not-found-error-cant-resolve-html-webpack-plugin-lib-loader-j-s).

**Comment by deebloo** [comment]
does heroku keep a cache you can clear? I know Ive had issues with Circle CI when updating dependencies and have had to rerun builds with no cache. Not sure if something like that is applicable here.

**Comment by JanStureNielsen** [comment]
@deebloo -- it does; to clear Heroku's cache:
```
heroku config:set NODEMODULESCACHE=false
git commit -am 'rebuild' --allow-empty
git push heroku master
heroku config:unset NODEMODULESCACHE
```
But I get the same strange error: `Error: Can't resolve 'html-webpack-plugin/lib/loader.j  s'`

**Comment by justinoverton** [comment]
I found the stackoverflow question and eventually found this entry also. I setup a new c9.io environment to try out angular2, and I got this issue on a fresh install and a new application using "ng new"

The output of "ng build" and the output of other commands for various version info is all in this gist:
https://gist.github.com/justinoverton/edcc294b95430f8ccfbb668e96e829f5

**Comment by JanStureNielsen** [comment]
The initial error is now resolved with the latest ``html-webpack-plugin@2.28.0`; but the next errors remain:
```
remote:        ERROR in multi ./src/main.ts
remote:        Module not found: Error: Can't resolve '@ngtools/webpack' in '/tmp/build_6a557f153da6022c883b49ad04b6209f'
remote:        @ multi ./src/main.ts
remote:        
remote:        ERROR in multi ./src/app/assets/css/main.scss
remote:        Module not found: Error: Can't resolve 'style-loader' in '/tmp/build_6a557f153da6022c883b49ad04b6209f'
remote:        @ multi ./src/app/assets/css/main.scss
```
Any ideas how I can configure the generated webpack configuration file to address these errors now?

**Comment by JanStureNielsen** [comment]
In my case, after moving my build dependencies from "devDep" to "dep", I can successfully build and deploy to Heroku. Thanks for the help.

**Comment by bofcarbon1** [comment]
I am trying to use angular2-cli on Cloud 9 as well. Ran into a similar problem But the Cloud 9 workspace makes things confusing. It generates package.json and a working sample if you select node.js. Since the instructions say to create a new project you end up creating this under the root which really should be the project directory. I'm wondering if these not found issues are the result of having to have dependencies under a directory of the project folder created when the new project command is given after the npm install -g anglular-cli command. This just infuriates me to no end. Not to mention that Cloud 9 be default uses a very old version of node.js. I'm upgrading to node 6.5.0. npm is like 3.10.3 which seems ok. I've read notes on versioning angular2 but have to look for that upgrade command if one exists. First I'm going to just try to add a project under the workspace even thought it will be like having a project in a project but if it works screw it. Getting tired of this.

**Comment by justinoverton** [comment]
@bofcarbon1 
With cloud9 what I did is as follows to sovle the issue, at least the steps I recall:

```
nvm install v6.5.0
nvm use v6.5.0

cd path_to_angular2_app
rm -rf node_modules

npm cache clean
#I may have uninstalled angular-cli and reinstalled it but I don't remember
npm uninstall -g angular-cli
npm install -g angular-cli

npm install
ng build
```

**Comment by justinoverton** [comment]
@bofcarbon1 

Also,

I created the angular project in a subdirectory of workspace.

Then I updated the server.js to serve static files from the dist directory of the angular app.

```
var http = require('http');
var path = require('path');

var async = require('async');
var socketio = require('socket.io');
var express = require('express');

//
// ## SimpleServer `SimpleServer(obj)`
//
// Creates a new instance of SimpleServer with the following options:
//  * `port` - The HTTP port to listen on. If `process.env.PORT` is set, _it overrides this value_.
//
var router = express();
var server = http.createServer(router);

router.use(express.static(path.resolve(__dirname, 'angular2appdir/dist')));

server.listen(process.env.PORT || 3000, process.env.IP || "0.0.0.0", function(){
  var addr = server.address();
  console.log("listening at", addr.address + ":" + addr.port);
});
```

I start the server with `npm start --host 0.0.0.0 --port 8080` which will make your application available at https://{workspacename}-{username}.c9users.io/

For me, using `ng serve ` would always eventually crash my workspace, but I have the minimum CPU and memory.

**Comment by bofcarbon1** [comment]
@justinoverton 

Appreciate the info.

I think in the future I am going to create a new work space selecting an empty option (if there is one).
I've been choosing node.js which constructs a default project. Then I follow the angular2-cli install and realize I am scaffolding on top of an already scaffolded workspace/project. I deleted the previous work space and started over last night. This time I used the node.js option but called it webresumeproject. I upgraded node.js to 6.5.0 then ran the angular2-sci install then created a new project webresume. Then changed directory to the webresume project then ran ng serve. So that actually got me something that resulted in a web workers are valid message. Although I was not able verify that a node.js server started an application that had a reachable URL. But I know that we have to use 8080,81 or 82 ports which I always have to adjust from any scaffolded project. So now I have to read up on where angular2-cli places the typescript files that we use to configure and start the server with the appropriate Cloud 9 ports and URL. And hey this stuff is free and open source so as much as I pull my hairs out and get frustrated its better than not having it.   Thanks
